File: prog5/Dictionary.c
//Dictionary.c
//Aaron Hom
//awhom
//pa5
//A Dictionary ADT that uses a hash table

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"
const int tableSize = 101;

// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
  char* key;
  char* val;
  struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* k, char* v) {
  Node P = malloc(sizeof(NodeObj));
  assert(P != NULL);
  P->key = k;
  P->val = v;
  P->next = NULL;
  return (P);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pP){
  if(pP != NULL && *pP != NULL){
    free(*pP);
    *pP = NULL;
  }
}

//ListObj
typedef struct ListObj{
  Node head;
} ListObj;

//List
typedef ListObj* List;

//newList()
//constructor for Node type
List newList(void){
  List L = malloc(sizeof(ListObj));
  assert(L != NULL);
  L->head = NULL;
  return L;
}

void freeList(List* pL){
  if(pL != NULL && *pL != NULL){
  free(*pL);
  *pL = NULL;
  }
}

//DictionaryObj
typedef struct DictionaryObj{
  int numItems;
  List table;
} DictionaryObj;

//hash functions --------------------------------------------------------------

//rotateLeft()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
  int sizeInBits = 8*sizeof(unsigned int);
  shift = shift & (sizeInBits - 1);
  if ( shift == 0 )
    return value;
  return (value << shift) | (value >> (sizeInBits - shift));
}

// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) { 
  unsigned int result = 0xBAE86554;
  while (*input) { 
    result ^= *input++;
    result = rotate_left(result, 5);
  }
  return result;
}

// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char* key){
  return pre_hash(key)%tableSize;
}

// private helper functions --------------------------------------------------

//findKey()
//returns the Node containing the key k in the tree subrooted at P, or returns NULL if no key exists
Node findKey(Node P, char* k){
  while(P != NULL){
    if(strcmp(P->key, k) == 0){
      break;
    } P = P->next;
  } return P;
}

//deleteAll()
//deletes all Nodes in the subtree rooted at P

  void deleteAll(Node P){
    if(P != NULL){
      deleteAll(P->next);
      freeNode(&P);
     }
  }
// public functions -----------------------------------------------------------

// Dictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){
  Dictionary D = malloc(sizeof(DictionaryObj));
  assert(D != NULL);
  D->table = calloc(tableSize, sizeof(ListObj));
  D->numItems = 0;
  return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
  if(pD != NULL && *pD != NULL){
    if(!isEmpty(*pD)) makeEmpty(*pD);
    free(*pD);
    *pD = NULL;
  }
}

// isEmpty()
// returns 1 (true) if D is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
  if(D == NULL){
    fprintf(stderr, "Dictionary Error: calling isEmpty() on unfound Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  return(D->numItems == 0);
}

//size()
//returns number of key/val (k, v) pairs in Dictionary D
//pre: none
int size(Dictionary D){
  return D->numItems;
}

// lookup()
// returns value that the pair key/val is in, and returns NULL it doesn't exist
// pre: none
char* lookup(Dictionary D, char* k){
  int tabIndex;
  Node P;
  List L;
  tabIndex = hash(k);
  L = &D->table[tabIndex];
  P = findKey(L->head, k);
  if(P == NULL)
    return NULL;
  else
    return P->val;
}

// insert()
//inserts a new key/val pair (k, v) into Dictionary D
// pre: key does not already exist
void insert(Dictionary D, char* k, char* v){
  int tabIndex;
  Node P;
  List L;
  tabIndex = hash(k);
  P = newNode(k, v);
  L = &D->table[tabIndex];
  if(findKey(L->head, k) != NULL){
    fprintf(stderr, "Dictionary Error: cannot insert a key that already exists\n");
    exit(EXIT_FAILURE);
  }
    P->next = L->head;
    L->head = P;
    P = NULL;
    D->numItems++;
}

//delete()
// deletes the pair assoicated with the key x
// pre: !isEmpty(D)
void delete(Dictionary D, char* k){
  if(D == NULL){
    fprintf(stderr, "Dictionary Error: cannot delete non existent key\n");
    exit(EXIT_FAILURE);
  }
  int tabIndex;
  Node P;
  List L;
  tabIndex = hash(k);
  L = &D->table[tabIndex];
  P = L->head;
  if(findKey(L->head, k) == L->head){
    P = L->head;
    L->head = L->head->next;
    freeNode(&P);
  }else{
    P = findKey(L->head, k);
    Node keep = L->head;
    Node temp = L->head->next;
    while(temp != P){
      temp = temp->next;
      keep = keep->next;
    }
    keep->next = P->next;
    P->next = NULL;
  }
  D->numItems--;
  freeNode(&P);
}

// makeEmpty()
// resets Dictionary D to an empty state
// pre: none
void makeEmpty(Dictionary D){
  List L;
  if(D->numItems == 0){
    fprintf(stderr, "Dictionary Error: calling makeEmpty on empty Dictionary\n");
  }
  if(D == NULL){
    fprintf(stderr, "Dictionary Error: calling makeEmpty on unfound Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  for(int i = 0; i < tableSize; i++){
    L = &D->table[i];
    deleteAll(L->head);
    free(&D->table[i]);
    }
  D->table = NULL;
  D->numItems = 0;
}

// printDictionary()
// prints a text representation of Dictionary A to the file pointed to by out
// pre: none
void printDictionary(FILE* out, Dictionary D){
  Node P;
  List L;
  if(D == NULL){
    fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  for(int i = 0; i < tableSize; i++){
    L = &D->table[i];
    P = L->head;
    while(P != NULL){
    fprintf(out, "%s %s \n", P->key, P->val);
    P = P->next;
    }
  }
}

File: lab7/Dictionary.java
//Dictionary.java
//Aaron Hom
//awhom
//lab7
//Translating the C version of binary search tree to java

// private types and functions ------------------------------------------------

public class Dictionary implements DictionaryInterface{

    //private inner node class
    private class Node{
	Pair item;
        Node left;
        Node right;

	Node(Pair p){
	    item = p;
	    left = null;
	    right = null;
	}
    }

    //private inner pair class
    private class Pair{
	String key;
	String val;

	Pair(String a, String b){
	    key = a;
	    val = b;
	}
    }

    //fields for Dictionary class
    private Node root; //first node in list (like head)
    private int numItems; //number of items

    //Dictionary() constructor
    public Dictionary(){
	root = null;
	numItems = 0;
    }

    //private helper functions ----------------------------------------------------------
    
    //findKey
    //returns Node that has key k in subtree rooted at Q, returns null if no Node exists
    Node findKey(Node P, String target){
	if(P == null || P.item.key.equals(target)){
            return P;
	}
	if(P.item.key.compareToIgnoreCase(target) > 0){
            return findKey(P.left, target);
	}else{
            return findKey(P.right, target);
	}
    }

    //findParent()
    //pre: P != null
    //returns parent of P in subtree root, returns null if P = Q
    Node findParent(Node P, Node Q){
	Node R = null;
	if(P != Q){
	    R = Q;
	    while(R.left != P && R.right != P){
		if(P.item.key.compareToIgnoreCase(P.item.key) < 0){
		    R = R.left;
		}else{
		    R = R.right;
		}
	    }
	}
	return R;
    }

    //findLeftmost()
    //returns leftmost Node in subtree rooted P, returns null is P is empty
    Node findLeftmost(Node P){
	Node Q = P;
	if(Q != null) for( ; Q.left != null; Q = Q.left);
	    return Q;
    }

     //printInOrder()
     //prints the key, val pairs from the subtree rooted at P in ascending order
     void printInOrder(Node P){
         if(P != null){
	     printInOrder(P.left);
	     System.out.println(P.item.key +" "+ P.item.val);
	     printInOrder(P.right);
	 }
     }

     //deleteAll()
     //deletes all Nodes in the subtree rooted at P
     void deleteAll(Node P){
	 if(P != null){
	     deleteAll(P.left);
	     deleteAll(P.right);
	 }
     }
       
     //ADT Operations --------------------------------------------------------------------
	
     //isEmpty()
     //pre: none
     //post: returns true if Dictionary is empty, false otherwise
     public boolean isEmpty(){
         return(numItems == 0);
     }

     //size()
     //pre: none
     //returns size of Dictionary
     public int size(){
         return numItems;
     } 

     //lookup
     //pre: none
     //post: returns val associated key, null if key doesn't exist
     public String lookup(String key){
	 Node P;
	 P = findKey(root, key);
	 return(P == null ? null : P.item.val);
     }

     //insert()
     //inserts new key, val pair into Dictionary
     //pre: lookup(key) == null
     //post: !isEmpty(), size++
     public void insert(String key, String val) throws DuplicateKeyException{
	 Node P, Q, R;
	 if(findKey(root, key) != null){
	     throw new DuplicateKeyException("Dictionary error: cannot insert already existing key");
	 }
	 P = new Node(new Pair(key, val));
	 Q = root;
	 R = null;
	 while(Q != null){
	     R = Q;
	     if(Q.item.key.compareToIgnoreCase(key) > 0){
		 Q = Q.left;
	     }else{
		 Q = Q.right;
	     }
	 }
         if(R == null){
	     root = P;
	 }else if(R.item.key.compareToIgnoreCase(key) > 0){
	     R.left = P;
	 }else{
	     R.right = P;
	 }
	 numItems++;
     }

     //delete()
     //deletes key and associated val
     //pre:lookup(key) != null
     public void delete(String key) throws KeyNotFoundException{
	 Node P, Q, R;
	 if(findKey(root, key) == null){
	     throw new KeyNotFoundException("Dictionary error: cannot delete non existent key");
	 }
	 P = findKey(root, key);
	 if(P.left == null && P.right == null){
	     if(P == root){
		 root = null;
	     }else{
		 Q = findParent(P, root);
		 if(Q.right == P) Q.right = null;
		 else Q.left = null;
	     }
	 }else if(P.right == null){
	     if(P == root){
		 root = P.left;
	     }else{
		 Q = findParent(P, root);
		 if(Q.right == P) Q.right = P.left;
		 else Q.left = P.left;
	     }
	 }else if(P.left == null){
	     if(P == root){
		 root = P.right;
	     }else{
		 Q = findParent(P, root);
		 if(Q.right == P) Q.right = P.right;
		 else Q.left = P.right;
	     }
	 }else{  
	     R = findLeftmost(P.right);
	     P.item.key = R.item.key;
	     P.item.val = R.item.val;
	     Q = findParent(R, P);
	     if(Q.right == R) Q.right = R.right;
	     else Q.left = R.right;
	 }
	 numItems--;
     }

     //makeEmpty()
     //pre: none
     public void makeEmpty(){
         deleteAll(root);
	 root = null;
	 numItems = 0;
     }

     //toString()
     //overrides Object's toString() method
     public String toString(){
         printInOrder(root);
	 return "";
     }
}
